File: app/src/main/java/com/keylesspalace/tusky/util/SpanUtils.java
/* Copyright 2017 Andrew Dawson
 *
 * This file is a part of Tusky.
 *
 * This program is free software; you can redistribute it and/or modify it under the terms of the
 * GNU General Public License as published by the Free Software Foundation; either version 3 of the
 * License, or (at your option) any later version.
 *
 * Tusky is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without even
 * the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General
 * Public License for more details.
 *
 * You should have received a copy of the GNU General Public License along with Tusky; if not,
 * see <http://www.gnu.org/licenses>. */

package com.keylesspalace.tusky.util;

import android.text.Spannable;
import android.text.Spanned;
import android.text.style.ForegroundColorSpan;

import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SpanUtils {
    /**
     * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/tag.rb">
     *     Tag#HASHTAG_RE</a>.
     */
    private static final String TAG_REGEX = "(?:^|[^/)\\w])#([\\w_]*[\\p{Alpha}_][\\w_]*)";
    private static Pattern TAG_PATTERN = Pattern.compile(TAG_REGEX, Pattern.CASE_INSENSITIVE);
    /**
     * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/account.rb">
     *     Account#MENTION_RE</a>
     */
    private static final String MENTION_REGEX =
            "(?:^|[^/[:word:]])@([a-z0-9_]+(?:@[a-z0-9\\.\\-]+[a-z0-9]+)?)";
    private static Pattern MENTION_PATTERN =
            Pattern.compile(MENTION_REGEX, Pattern.CASE_INSENSITIVE);

    private static final String URL_REGEX = "(?:(^|\\b)https?:[^\\s]+)";
    private static Pattern URL_PATTERN = Pattern.compile(URL_REGEX, Pattern.CASE_INSENSITIVE);

    private static class FindCharsResult {
        int charIndex;
        int stringIndex;
        boolean atWordBreak;

        FindCharsResult() {
            charIndex = -1;
            stringIndex = -1;
            atWordBreak = false;
        }
    }

    private static FindCharsResult findChars(String string, int fromIndex, char[] chars) {
        FindCharsResult result = new FindCharsResult();
        final int length = string.length();
        for (int i = fromIndex; i < length; i++) {
            char c = string.charAt(i);
            for (int j = 0; j < chars.length; j++) {
                if (chars[j] == c) {
                    result.charIndex = j;
                    result.stringIndex = i;
                    return result;
                }
            }
        }
        return result;
    }

    private static FindCharsResult findStart(String string, int fromIndex, char[] chars) {
        FindCharsResult found = findChars(string, fromIndex, chars);
        int i = found.stringIndex;
        if (i < 0) {
            return new FindCharsResult();
        } else if (i == 0 || i >= 1 && Character.isWhitespace(string.codePointBefore(i))) {
            found.atWordBreak = true;
        }
        return found;
    }

    private static int findEndOfPattern(String string, int fromIndex, Pattern pattern) {
        Matcher matcher = pattern.matcher(string);
        return matcher.find(fromIndex) ? matcher.end() : -1;
    }

    /** Takes text containing mentions and hashtags and makes them the given colour. */
    public static void highlightSpans(Spannable text, int colour) {
        // Strip all existing colour spans.
        int n = text.length();
        ForegroundColorSpan[] oldSpans = text.getSpans(0, n, ForegroundColorSpan.class);
        for (int i = oldSpans.length - 1; i >= 0; i--) {
            text.removeSpan(oldSpans[i]);
        }
        // Colour the mentions and hashtags.
        String string = text.toString();
        int start;
        int end = 0;
        while (end < n) {
            char[] chars = { ':', '#', '@' };
            FindCharsResult found = findStart(string, end, chars);
            start = found.stringIndex;
            if (start < 0) {
                break;
            }
            if (found.charIndex == 0) {
                int aNewStart = Math.max(0, start - 5);
                end = findEndOfPattern(string, aNewStart, URL_PATTERN);
                if (end >= 0) {
                    text.setSpan(new CustomURLSpan(string.substring(aNewStart, end)), aNewStart, end, Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                }
            } else if (found.charIndex == 1 && found.atWordBreak) {
                end = findEndOfPattern(string, Math.max(0, start - 1), TAG_PATTERN);
                if (end >= 0) {
                    text.setSpan(new ForegroundColorSpan(colour), start, end, Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                }
            } else if (found.charIndex == 2 && found.atWordBreak) {
                end = findEndOfPattern(string, Math.max(0, start - 1), MENTION_PATTERN);
                if (end >= 0) {
                    text.setSpan(new ForegroundColorSpan(colour), start, end, Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
                }
            } else {
                break;
            }
            if (end < 0) {
                break;
            }
        }
    }
}

File: app/src/main/java/com/keylesspalace/tusky/util/SpanUtils.kt
package com.keylesspalace.tusky.util

import android.content.Context
import android.text.Spannable
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.CharacterStyle
import android.text.style.DynamicDrawableSpan
import android.text.style.ForegroundColorSpan
import android.text.style.ImageSpan
import android.text.style.URLSpan
import androidx.appcompat.content.res.AppCompatResources
import com.keylesspalace.tusky.R
import com.keylesspalace.tusky.util.twittertext.Regex
import java.util.regex.Pattern

/**
 * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/tag.rb">
 *     Tag#HASHTAG_RE</a>.
 */
private const val HASHTAG_SEPARATORS = "_\\u00B7\\u30FB\\u200c"
internal const val TAG_PATTERN_STRING = "(?<![=/)\\p{Alnum}])(#(([\\w_][\\w$HASHTAG_SEPARATORS]*[\\p{Alpha}$HASHTAG_SEPARATORS][\\w$HASHTAG_SEPARATORS]*[\\w_])|([\\w_]*[\\p{Alpha}][\\w_]*)))"
private val TAG_PATTERN = TAG_PATTERN_STRING.toPattern(Pattern.CASE_INSENSITIVE)

/**
 * @see <a href="https://github.com/tootsuite/mastodon/blob/master/app/models/account.rb">
 *     Account#MENTION_RE</a>
 */
private const val USERNAME_PATTERN_STRING = "[a-z0-9_]+([a-z0-9_.-]+[a-z0-9_]+)?"
internal const val MENTION_PATTERN_STRING = "(?<![=/\\w])(@($USERNAME_PATTERN_STRING)(?:@[\\w.-]+[\\w]+)?)"
private val MENTION_PATTERN = MENTION_PATTERN_STRING.toPattern(Pattern.CASE_INSENSITIVE)

private val VALID_URL_PATTERN = Regex.VALID_URL_PATTERN_STRING.toPattern(Pattern.CASE_INSENSITIVE)

private val spanClasses = listOf(ForegroundColorSpan::class.java, URLSpan::class.java)

// url must come first, it may contain the other patterns
val defaultFinders = listOf(
    PatternFinder("http", FoundMatchType.HTTP_URL, VALID_URL_PATTERN),
    PatternFinder("#", FoundMatchType.TAG, TAG_PATTERN),
    PatternFinder("@", FoundMatchType.MENTION, MENTION_PATTERN)
)

enum class FoundMatchType {
    HTTP_URL,
    HTTPS_URL,
    TAG,
    MENTION
}

class PatternFinder(
    val searchString: String,
    val type: FoundMatchType,
    val pattern: Pattern
)

/**
 * Takes text containing mentions and hashtags and urls and makes them the given colour.
 * @param finders The finders to use. This is here so they can be overridden from unit tests.
 */
fun Spannable.highlightSpans(colour: Int, finders: List<PatternFinder> = defaultFinders) {
    // Strip all existing colour spans.
    for (spanClass in spanClasses) {
        clearSpans(spanClass)
    }

    for (finder in finders) {
        // before running the regular expression, check if there is even a chance of it finding something
        if (this.contains(finder.searchString, ignoreCase = true)) {
            val matcher = finder.pattern.matcher(this)

            while (matcher.find()) {
                // we found a match
                val start = matcher.start(1)

                val end = matcher.end(1)

                // only add a span if there is no other one yet (e.g. the #anchor part of an url might match as hashtag, but must be ignored)
                if (this.getSpans(start, end, URLSpan::class.java).isEmpty()) {
                    this.setSpan(
                        getSpan(finder.type, this, colour, start, end),
                        start,
                        end,
                        Spanned.SPAN_INCLUSIVE_EXCLUSIVE
                    )
                }
            }
        }
    }
}

private fun <T> Spannable.clearSpans(spanClass: Class<T>) {
    for (span in getSpans(0, length, spanClass)) {
        removeSpan(span)
    }
}

private val iconNameMapping: Map<String, Int> = mapOf(
    "{{home}}" to R.drawable.ic_home_24dp,
    "{{mail}}" to R.drawable.ic_mail_24dp,
    "{{group}}" to R.drawable.ic_group_24dp,
    "{{search}}" to R.drawable.ic_search_24dp,
    "{{manage_accounts}}" to R.drawable.ic_manage_accounts_24dp,
    "{{chevron_right}}" to R.drawable.ic_chevron_right_24dp,
)

/**
 * Replaces text of the form {{icon_name}} with their spanned counterparts (ImageSpan). Supported icon names are above.
 */
fun addDrawables(text: CharSequence, color: Int, size: Int, context: Context): Spannable {
    val builder = SpannableStringBuilder(text)

    iconNameMapping.forEach { iconName, icon ->
        var index = 0
        while (index < text.length - iconName.length && index != -1) {
            index = text.indexOf(iconName, index)

            if (index != -1) {
                val drawable = AppCompatResources.getDrawable(context, icon)!!
                drawable.setBounds(0, 0, size, size)
                drawable.setTint(color)
                builder.setSpan(ImageSpan(drawable, DynamicDrawableSpan.ALIGN_CENTER), index, index + iconName.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
                index += iconName.length
            }
        }
    }

    return builder
}

private fun getSpan(
    matchType: FoundMatchType,
    string: CharSequence,
    colour: Int,
    start: Int,
    end: Int
): CharacterStyle {
    return when (matchType) {
        FoundMatchType.HTTP_URL, FoundMatchType.HTTPS_URL -> NoUnderlineURLSpan(string.substring(start, end))
        FoundMatchType.MENTION -> MentionSpan(string.substring(start, end))
        else -> ForegroundColorSpan(colour)
    }
}
